ddtrace/opentelemetry/span_test.go

```diff
@@ -22,63 +24,248 @@ import (
+		_, sp := tr.Start(context.Background(), test.trueName)
+		sp.SetStatus(codes.Error, test.trueErrorMsg)
+		for k, v := range test.trueAttributes {
+			sp.SetAttributes(attribute.String(k, v))
+		}
+		assert.True(sp.IsRecording())
+		sp.End()
+		assert.False(sp.IsRecording())
+		// following operations should not be able to modify the Span
+		sp.SetName(test.trueName)
+		sp.SetStatus(test.falseError, test.falseErrorMsg)
+		for k, v := range test.trueAttributes {
+			sp.SetAttributes(attribute.String(k, v))
+			sp.SetAttributes(attribute.String(k, v))
+		}
 
-	_, sp := tr.Start(context.Background(), testData.oldOp)
-	for _, tag := range testData.tags {
+		tracer.Flush()
+		payload, err := waitForPayload(ctx, payloads)
+		if err != nil {
+			t.Log(err)
+			t.Fail()
+		}
+		assert.Contains(payload, test.trueErrorMsg)
+		assert.NotContains(payload, test.falseErrorMsg)
+		assert.Contains(payload, test.trueName)
+		assert.NotContains(payload, test.falseName)
+		for k, v := range test.trueAttributes {
+			assert.Contains(payload, fmt.Sprintf("\"%s\":\"%s\"", k, v))
+		}
+		for k, v := range test.falseAttributes {
+			assert.NotContains(payload, fmt.Sprintf("\"%s\":\"%s\"", k, v))
+		}
+		cancel()
+	}
+}
+
+func TestSpanSetStatus(t *testing.T) {
+	assert := assert.New(t)
+	testData := []struct {
+		higherCode     codes.Code
+		higherCodeDesc string
+		lowerCode      codes.Code
+		lowerCodeDesc  string
+	}{
+		{
+			higherCode:     codes.Ok,
+			higherCodeDesc: "ok_description",
+			lowerCode:      codes.Error,
+			lowerCodeDesc:  "error_description",
+		},
+		{
+			higherCode:     codes.Error,
+			higherCodeDesc: "error_description",
+			lowerCode:      codes.Unset,
+			lowerCodeDesc:  "unset_description",
+		},
+	}
+
+	_, payloads, cleanup := mockTracerProvider(t)
+	tr := otel.Tracer("")
+	defer cleanup()
+
+	for _, test := range testData {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+
+		_, sp := tr.Start(context.Background(), "test")
+		sp.SetStatus(test.higherCode, test.higherCodeDesc)
+		sp.SetStatus(test.lowerCode, test.lowerCodeDesc)
+		sp.End()
+
+		tracer.Flush()
+		payload, err := waitForPayload(ctx, payloads)
+		if err != nil {
+			t.Log(err)
+			t.Fail()
+		}
+		if test.higherCode == codes.Error {
+			assert.Contains(payload, test.higherCodeDesc)
+		} else {
+			assert.NotContains(payload, test.higherCodeDesc)
+		}
+		assert.NotContains(payload, test.lowerCodeDesc)
+
+		_, sp = tr.Start(context.Background(), "test")
+		sp.SetStatus(test.lowerCode, test.lowerCodeDesc)
+		sp.SetStatus(test.higherCode, test.higherCodeDesc)
+		sp.End()
+
+		tracer.Flush()
+		payload, err = waitForPayload(ctx, payloads)
+		if err != nil {
+			t.Log(err)
+			t.Fail()
+		}
+		if test.higherCode == codes.Error {
+			assert.Contains(payload, test.higherCodeDesc)
+		} else {
+			assert.NotContains(payload, test.higherCodeDesc)
+		}
+		assert.NotContains(payload, test.lowerCodeDesc)
+		cancel()
```

Comment:
This code is doing almost exactly the same thing as the code above it, so that tells me we should probably just have a function. You can do that inline, with something like this:

```
testStatus := func() {
	sp.End()

	tracer.Flush()
	payload, err := waitForPayload(ctx, payloads)
	if err != nil {
		t.Fatal(err)
	}
	if test.higherCode == codes.Error {
		assert.Contains(payload, test.higherCodeDesc)
	} else {
		assert.NotContains(payload, test.higherCodeDesc)
	}
	assert.NotContains(payload, test.lowerCodeDesc)
}

_, sp := tr.Start(context.Background(), "test")
sp.SetStatus(test.higherCode, test.higherCodeDesc)
sp.SetStatus(test.lowerCode, test.lowerCodeDesc)
testStatus()

// Flip the order by which the status is set, but ensure that
// the code with the higher precedence is used.
_, sp := tr.Start(context.Background(), "test")
sp.SetStatus(test.lowerCode, test.lowerCodeDesc)
sp.SetStatus(test.higherCode, test.higherCodeDesc)
testStatus()

```